src/Stack/Setup.hs

```diff
@@ -140,6 +140,8 @@ data SetupOpts = SetupOpts
     , soptsUpgradeCabal :: !Bool
     -- ^ Upgrade the global Cabal library in the database to the newest
     -- version. Only works reliably with a stack-managed installation.
+    , soptsInstallCabal :: !(Maybe Version)
+    -- ^ Install a specific version of Cabal.
```

Comment:
Can you fuse the `soptsUpgradeCabal` and `soptsInstallCabal` fields into one? This should make it impossible to have nonsensical states where `soptsUpgradeCabal == True` and `soptsInstallCabal == Just <x>`.

Maybe it would even be good to avoid the additional `--install-cabal` flag and only add an optional argument to `--upgrade-cabal`?! That would help combating CLI inflation :). Not sure if it's possible with optparse-applicative though…